crates/ruma-events/src/room/message/captionable.rs

```diff
@@ -0,0 +1,44 @@
+//! Reusable methods for captioning media files.
+
+use crate::room::message::FormattedBody;
+
+/// A trait that provides caption methods for the relevant message event content types.
+pub trait Captionable {
+    /// The textual representation of the message.
+    fn body(&self) -> &str;
+    /// Formatted form of the message `body`.
+    fn formatted(&self) -> Option<&FormattedBody>;
+    /// The original filename of the uploaded file.
+    fn filename(&self) -> Option<&str>;
```

Comment:
This seems all kinds of wrong to me, but was the only way I could find to share the implementations of `caption` and `formatted_caption`. Happy to go a different route if a better one exists 🙂

Reply:
What I would do instead is make private free-standing `caption` and `formatted_caption` functions, and public inherent ones on all the relevant types that call the private ones. Alternatively copy-paste the implementation. Could use some combinators to make it pretty short.

Reply:
Updated in https://github.com/ruma/ruma/pull/1919/commits/70e8af99faa7d7d90a738361775021edf942d6a4, I think that's doing what you described?